src/core/lib/event_engine/ares_resolver.cc

```diff
@@ -0,0 +1,714 @@
+            ares_resolver, output, hostname_qa->port);
+        break;
+      }
+    }
+  }
+  ares_resolver->event_engine_->Run(
+      [callback = std::move(callback), result = std::move(result)]() mutable {
+        callback(std::move(result));
+      });
+}
+
+void AresResolver::OnSRVQueryDoneLocked(void* arg, int status, int /*timeouts*/,
+                                        unsigned char* abuf, int alen) {
+  std::unique_ptr<QueryArg> qa(static_cast<QueryArg*>(arg));
+  auto* ares_resolver = qa->ares_resolver;
+  auto nh = ares_resolver->callback_map_.extract(qa->callback_map_id);
+  GPR_ASSERT(!nh.empty());
+  GPR_ASSERT(
+      absl::holds_alternative<EventEngine::DNSResolver::LookupSRVCallback>(
+          nh.mapped()));
+  auto callback = absl::get<EventEngine::DNSResolver::LookupSRVCallback>(
+      std::move(nh.mapped()));
+  if (status != ARES_SUCCESS) {
+    std::string error_msg = absl::StrFormat(
+        "c-ares status is not ARES_SUCCESS qtype=SRV name=%s: %s",
+        qa->qname.c_str(), ares_strerror(status));
+    GRPC_ARES_RESOLVER_TRACE_LOG("request:%p on_srv_done_locked: %s",
+                                 ares_resolver, error_msg.c_str());
+    ares_resolver->event_engine_->Run(
+        [callback = std::move(callback),
+         status = AresStatusToAbslStatus(status, error_msg)]() mutable {
+          callback(status);
+        });
+    return;
+  }
+  struct ares_srv_reply* reply = nullptr;
+  const int parse_status = ares_parse_srv_reply(abuf, alen, &reply);
+  GRPC_ARES_RESOLVER_TRACE_LOG("request:%p ares_parse_srv_reply: %d",
+                               ares_resolver, parse_status);
+  std::vector<EventEngine::DNSResolver::SRVRecord> result;
+  if (parse_status == ARES_SUCCESS) {
+    for (struct ares_srv_reply* srv_it = reply; srv_it != nullptr;
+         srv_it = srv_it->next) {
+      EventEngine::DNSResolver::SRVRecord record;
+      record.host = srv_it->host;
+      record.port = srv_it->port;
+      record.priority = srv_it->priority;
+      record.weight = srv_it->weight;
+      result.push_back(std::move(record));
+    }
+  }
+  if (reply != nullptr) {
+    ares_free_data(reply);
+  }
+  ares_resolver->event_engine_->Run(
+      [callback = std::move(callback), result = std::move(result)]() mutable {
+        callback(std::move(result));
+      });
+}
+
+void AresResolver::OnTXTDoneLocked(void* arg, int status, int /*timeouts*/,
+                                   unsigned char* buf, int len) {
+  std::unique_ptr<QueryArg> qa(static_cast<QueryArg*>(arg));
+  auto* ares_resolver = qa->ares_resolver;
+  auto nh = ares_resolver->callback_map_.extract(qa->callback_map_id);
+  GPR_ASSERT(!nh.empty());
+  GPR_ASSERT(
+      absl::holds_alternative<EventEngine::DNSResolver::LookupTXTCallback>(
+          nh.mapped()));
+  auto callback = absl::get<EventEngine::DNSResolver::LookupTXTCallback>(
+      std::move(nh.mapped()));
+  struct ares_txt_ext* reply = nullptr;
+  int parse_status = ARES_SUCCESS;
```

Comment:
I think the logic here would be a bit clearer if we handled one status at a time.  We can avoid code duplication with a simple lambda:

```
auto fail = [&](absl::string_view prefix) {
  std::string error_message = absl::StrFormat("%s for %s: %s",
      prefix, qa->qname, ares_strerror(status));
  GRPC_ARES_RESOLVER_TRACE_LOG("request:%p on_txt_done_locked: %s",
                                 ares_resolver, error_msg.c_str());
  ares_resolver->event_engine_->Run(
      [callback = std::move(callback),
        status = AresStatusToAbslStatus(status, error_msg)]() mutable {
        callback(status);
      });
};
if (status != ARES_SUCCESS) {
  fail("TXT lookup failed");
  return;
}
GRPC_ARES_RESOLVER_TRACE_LOG(
        "request:%p on_txt_done_locked name=%s ARES_SUCCESS", ares_resolver,
        qa->qname.c_str());
status = ares_parse_txt_reply_ext(buf, len, &reply);
if (status != ARES_SUCCESS) {
  fail("parsing TXT result failed");
  return;
}
```

Reply:
Great point! No need for the `parse_status` local variable as well. Very neat!